txpool/tx_pool.go

```diff
@@ -456,8 +457,8 @@ func (p *TxPool) wash(headSummary *chain.BlockSummary) (executables tx.Transacti
 
 	for _, obj := range executableObjs {
 		executables = append(executables, obj.Transaction)
-		if !obj.executable || obj.localSubmitted {
```

Comment:
This is a curious piece of code: here we're checking the executable value, but we know that it should be true, since it was computed previously [here](https://github.com/vechain/thor/blob/master/txpool/tx_pool.go#L398) and put in `executableObjs` slice accordingly. The code in between doesn't seem to change the executable value. Nonetheless, we check it again. This makes me think there is a concurrent function that could potentially change it (during the subscriptions tests a data race was detected when adding a tx to the mempool). Here is why this PR introduces a sync read/write of executable

Reply:
There should be no race, `add` creates the txObj and is added to the `p.all` which has a lock inside the add function. And in the `wash` func, the list of executableObj is copied from `p.all.ToTxObjects()`, also a lock inside, so no need for the locks inside `txObj`.

Reply:
As for the late updated executable property, it is intended to do so. The list can be trimmed due to the pool limit, so update the executable property after all work has been done.

Reply:
I noticed the concurrent access in this [test run](https://github.com/vechain/thor/actions/runs/8479934212/job/23234781081#step:5:149)

```
==================
WARNING: DATA RACE
Read at 0x00c00602a258 by goroutine 103:
  github.com/vechain/thor/v2/txpool.(*TxPool).wash()
      /home/runner/work/thor/thor/txpool/tx_pool.go:459 +0x20e4
  github.com/vechain/thor/v2/txpool.(*TxPool).housekeeping()
      /home/runner/work/thor/thor/txpool/tx_pool.go:121 +0x5d9
  github.com/vechain/thor/v2/txpool.(*TxPool).housekeeping-fm()
      <autogenerated>:1 +0x33
  github.com/vechain/thor/v2/co.(*Goes).Go.func1()
      /home/runner/work/thor/thor/co/goes.go:22 +0x86

Previous write at 0x00c00602a258 by goroutine 92:
  github.com/vechain/thor/v2/txpool.(*TxPool).add()
      /home/runner/work/thor/thor/txpool/tx_pool.go:259 +0xb9a
  github.com/vechain/thor/v2/txpool.(*TxPool).AddLocal()
      /home/runner/work/thor/thor/txpool/tx_pool.go:[29](https://github.com/vechain/thor/actions/runs/8479934212/job/23338972173#step:5:30)1 +0x3ec
  github.com/vechain/thor/v2/api/subscriptions.TestPendingTx_DispatchLoop()
      /home/runner/work/thor/thor/api/subscriptions/pending_tx_test.go:81 +0x3d5
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1648 +0x44

Goroutine 103 (running) created at:
  github.com/vechain/thor/v2/co.(*Goes).Go()
      /home/runner/work/thor/thor/co/goes.go:20 +0xe4
  github.com/vechain/thor/v2/txpool.New()
      /home/runner/work/thor/thor/txpool/tx_pool.go:82 +0x44b
  github.com/vechain/thor/v2/api/subscriptions.TestPendingTx_DispatchLoop()
      /home/runner/work/thor/thor/api/subscriptions/pending_tx_test.go:61 +0x1a7
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1648 +0x[44](https://github.com/vechain/thor/actions/runs/8479934212/job/23338972173#step:5:45)

Goroutine 92 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1648 +0x8[45](https://github.com/vechain/thor/actions/runs/8479934212/job/23338972173#step:5:46)
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:2054 +0x84
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.runTests()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:2052 +0x8ad
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1925 +0xcd7
  main.main()
      _testmain.go:113 +0x2e4
==================
```

Reply:
Changing the above code in `add` to:

``` go
		txObj.executable = executable
		if err := p.all.Add(txObj, p.options.LimitPerAccount); err != nil {
			return txRejectedError{err.Error()}
		}
```

Will solve the issue.

Reply:
Ok, I've ran multiple tests and I'm not getting the race condition anymore. I'm still wondering why moving that line above fixes the issue: couldn't the wash procedure still read the executable field concurrently while the add is changing its value? There is no active lock when those operations happen